File: src/main/java/seremis/geninfusion/tileentity/GITile.java
package seremis.geninfusion.tileentity;

import cpw.mods.fml.common.network.NetworkRegistry.TargetPoint;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.Packet;
import net.minecraft.network.play.server.S35PacketUpdateTileEntity;
import net.minecraft.tileentity.TileEntity;
import seremis.geninfusion.GeneticInfusion;
import seremis.geninfusion.core.proxy.CommonProxy;
import seremis.geninfusion.network.packet.PacketTileData;

public class GITile extends TileEntity {

    private int teDirection;

    public int getDirection() {
        return teDirection;
    }

    public void setDirection(int direction) {
        this.teDirection = direction;
        worldObj.addBlockEvent(xCoord, yCoord, zCoord, getBlockType(), 1, direction);
    }

    public boolean isUseableByPlayer(EntityPlayer player) {
        return true;
    }

    @Override
    public boolean receiveClientEvent(int eventId, int variable) {
        if(eventId == 1) {
            teDirection = variable;
            worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
            return true;
        } else {
            return super.receiveClientEvent(eventId, variable);
        }
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);
        teDirection = compound.getInteger("teDirection");
    }

    @Override
    public void writeToNBT(NBTTagCompound compound) {
        super.writeToNBT(compound);
        compound.setInteger("teDirection", teDirection);
    }

    @Override
    public void onDataPacket(NetworkManager net, S35PacketUpdateTileEntity packet) {
        readFromNBT(packet.func_148857_g());
    }

    @Override
    public Packet getDescriptionPacket() {
        NBTTagCompound compound = new NBTTagCompound();
        writeToNBT(compound);
        return new S35PacketUpdateTileEntity(this.xCoord, this.yCoord, this.zCoord, 1, compound);
    }

    public void sendTileDataToServer(int id, byte[] data) {
        if(CommonProxy.instance.isRenderWorld(worldObj)) {
            GeneticInfusion.packetPipeline.sendToServer(new PacketTileData(data, id, this.xCoord, this.yCoord, this.zCoord));
        } else {
            setTileDataFromClient(id, data);
        }
    }

    public void setTileDataFromServer(int id, byte[] data) {

    }

    public void sendTileDataToClient(int id, byte[] data) {
        if(CommonProxy.instance.isServerWorld(worldObj)) {
            GeneticInfusion.packetPipeline.sendToAllAround(new PacketTileData(data, id, xCoord, yCoord, zCoord), new TargetPoint(worldObj.provider.dimensionId, xCoord, yCoord, zCoord, 128));
        } else {
            setTileDataFromServer(id, data);
        }
    }

    public void setTileDataFromClient(int id, byte[] data) {

    }
}

File: src/main/scala/seremis/geninfusion/tileentity/GITile.scala
package seremis.geninfusion.tileentity

import cpw.mods.fml.common.network.NetworkRegistry.TargetPoint
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.play.server.S35PacketUpdateTileEntity
import net.minecraft.network.{NetworkManager, Packet}
import net.minecraft.tileentity.TileEntity
import seremis.geninfusion.GeneticInfusion
import seremis.geninfusion.network.ModPackets
import seremis.geninfusion.network.packet.PacketTileData

trait GITile extends TileEntity {

    private var direction: Int = _

    def getDirection: Int = direction

    def setDirection(direction: Int) {
        this.direction = direction
        getWorld.addBlockEvent(xCoord, yCoord, zCoord, getBlockType, 1, direction)
    }

    def isUseableByPlayer(player: EntityPlayer): Boolean = true

    override def receiveClientEvent(eventId: Int, variable: Int): Boolean = {
        if (eventId == 1) {
            direction = variable
            getWorld.markBlockForUpdate(xCoord, yCoord, zCoord)
            true
        } else {
            super.receiveClientEvent(eventId, variable)
        }
    }

    override def readFromNBT(compound: NBTTagCompound) {
        super.readFromNBT(compound)
        direction = compound.getInteger("direction")
    }

    override def writeToNBT(compound: NBTTagCompound) {
        super.writeToNBT(compound)
        compound.setInteger("direction", direction)
    }

    @SideOnly(Side.CLIENT)
    override def onDataPacket(net: NetworkManager, packet: S35PacketUpdateTileEntity) {
        readFromNBT(packet.getNbtCompound)
    }

    override def getDescriptionPacket: Packet = {
        val compound = new NBTTagCompound()
        writeToNBT(compound)
        new S35PacketUpdateTileEntity(this.xCoord, this.yCoord, this.zCoord, 1, compound)
    }

    def sendTileDataToServer(id: Int, data: Array[Byte]) {
        if (GeneticInfusion.commonProxy.isRenderWorld(getWorld)) {
            ModPackets.wrapper.sendToServer(new PacketTileData(data, id, this.xCoord, this.yCoord, this.zCoord))
        } else {
            setTileDataFromClient(id, data)
        }
    }

    def setTileDataFromServer(id: Int, data: Array[Byte]) {}

    def sendTileDataToClient(id: Int, data: Array[Byte]) {
        if (GeneticInfusion.commonProxy.isServerWorld(getWorld)) {
            ModPackets.wrapper.sendToAllAround(new PacketTileData(data, id, xCoord, yCoord, zCoord), new TargetPoint(getWorld.provider.dimensionId, xCoord, yCoord, zCoord, 128))
        } else {
            setTileDataFromServer(id, data)
        }
    }

    def setTileDataFromClient(id: Int, data: Array[Byte]) {}
}
